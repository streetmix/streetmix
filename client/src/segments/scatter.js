import seedrandom from 'seedrandom'
import { images } from '../app/load_resources'
import { maxBy } from '../util/maxBy'
import { drawSegmentImage } from './view'
import { getSpriteDef } from './info'
import { TILE_SIZE, TILE_SIZE_ACTUAL } from './constants'

const DEFAULT_SELECTION_WEIGHT = 50
const DEFAULT_SCATTER_SPACING_MIN = 0
const DEFAULT_SCATTER_SPACING_MAX = 1 // in meters

/**
 * Given a pool of entities, with defined widths, get the maximum width value
 * This number is used during rendering to adjust sprites for center alignment
 *
 * @param {Array} pool - of objects to draw from
 * @returns {Number} - maximum street width in meters
 */
function getSpriteMaxWidth (pool) {
  return maxBy(pool, (s) => s.width).width
}

/**
 * Using the random generator function, perform a weighted random selection
 * from a pool of entities, then clone and return the object for that entity.
 *
 * @param {Array} pool - of entities to draw from. See the heredoc for
 *    getRandomObjects() for the expected data structure of entities.
 * @param {Function} randomGenerator - random generator function created
 *    from randSeed. Do not recreate random generator function within a loop,
 *    because that is not performant.
 */
function pickRandomEntityFromPool (pool, randomGenerator) {
  const totalWeight = pool.reduce((sum, entity) => {
    return sum + (entity.weight || DEFAULT_SELECTION_WEIGHT)
  }, 0)
  const randomNumber = Math.floor(randomGenerator() * totalWeight)

  let pickedEntityIndex
  let searchWeight = 0

  for (let i = 0; i < pool.length; i++) {
    const entity = pool[i]
    searchWeight += entity.weight || DEFAULT_SELECTION_WEIGHT

    if (randomNumber < searchWeight) {
      pickedEntityIndex = i
      break
    }
  }

  // Clone the entity and return it
  return Object.assign({}, pool[pickedEntityIndex])
}

/**
 * Given a pool of entities, and a few parameters, draw continuously from that
 * pool until it has fully populated the given segment width. Because it uses
 * segment width to know when to stop drawing, this function also does double
 * duty by setting distances between objects.
 *
 * This function is exported for unit testing only.
 *
 * @param {Array} pool - of entities to draw from
 *   Each entity is an object type with the following properties:
 *     - id {string} - a unique identifier. For drawing segments, this `id`
 *       corresponds to the sprite's "filename", but that association is not
 *       particularly meaningful for picking objects.
 *     - name {string} - (optional) human readable name of the entity
 *     - width {number} - the width that this entity occupies
 *     - weight {number} - (optional) affects the rarity of this entity. The
 *       higher the value, the higher the chance of it being selected, relative
 *       to others. The default weight value is 50. Lower numbers are rarer, higher
 *       numbers are more common.
 *     - disallowFirst {boolean} - (optional) if true, prevent this entity from
 *       being selected first in a list (reserved for unique or special entities
 *       that should not be the only object displayed if we only have space for
 *       one object.)
 * @param {Number} width - of segment to populate, in meters
 * @param {Number} randSeed - ensures a consistent sequence of objects across renders
 * @param {Number} minSpacing - minimum spacing between each object, in meters (controls density)
 * @param {Number} maxSpacing - maximt spacing between each object, in meters (controls density)
 * @param {Number} maxSpriteWidth - maximum sprite width in the pool (via `getSpriteMaxWidth()`)
 * @param {Number} spacingAdjustment - additional value to adjust spacing, in meters
 * @param {Number} padding - buffer zone at segment sides to avoid drawing in
 */
export function getRandomObjects (
  pool,
  maxWidth,
  randSeed = 9123984, // self defined randSeed if one is not provided
  minSpacing,
  maxSpacing,
  maxSpriteWidth,
  spacingAdjustment = 0,
  padding = 0
) {
  const randomGenerator = seedrandom(randSeed)

  const objects = []
  let runningWidth = 0
  let previousEntityId = null
  let currentEntityId = null
  let lastWidth = 0

  // Pick (draw) objects from the pool until we have at least one object, and
  // until we hit the maximum width given. `padding` is used to so that
  // we don't render objects too closely to the edge of a segment.
  while (objects.length === 0 || runningWidth < maxWidth - padding * 2) {
    let object

    // Special choosing logic only for larger pools.
    if (pool.length >= 4) {
      do {
        object = pickRandomEntityFromPool(pool, randomGenerator)
        currentEntityId = object.id
      } while (
        // Don't pick the same object twice in a row, and avoid picking
        // certain objects as the first object in a list. This is because
        // some objects are "special" and look strange when on their own,
        // but can diversify when included in a list.
        currentEntityId === previousEntityId ||
        (objects.length === 0 && object.disallowFirst === true)
      )
    } else {
      object = pickRandomEntityFromPool(pool, randomGenerator)
      currentEntityId = object.id
    }

    previousEntityId = currentEntityId
    object.left = runningWidth

    // Calculate the amount of space to allocate to this object,
    // creating space for placing the next object (if any).
    // First, take into account the object's defined sprite width.
    // Then add `minSpacing`, the minimum spacing between sprites.
    // Next, add a variable distance between sprites. This number
    // is randomly chosen between `minSpacing` and `maxSpacing.`
    // The `randomGenerator()` must be used to ensure that this random
    // number is consistent across renders.
    // Lastly, this is further tweaked by adding a `spacingAdjustment`
    // value, which can adjust overall density. In the case of people,
    // this value makes sprites slightly denser, to account for the
    // transparent space around each person sprite's defined width.
    lastWidth =
      object.width +
      minSpacing +
      randomGenerator() * (maxSpacing - minSpacing) +
      spacingAdjustment
    runningWidth += lastWidth

    objects.push(object)
  }

  // After exiting the loop, remove the space allocated for the next object,
  // by undoing the addition of `lastWidth` to `runningWidth`.
  // TODO: Refactor this behavior.
  const totalWidth = runningWidth - lastWidth

  // The left position of sprites are along the left edge.
  // This correction adjusts the actual left render position, based on the
  // width of the first and last sprites, to better balance centering the
  // sprites within the segment.
  const firstObjectCorrection = (maxSpriteWidth - objects[0].width) / 2
  const lastObjectCorrection =
    (maxSpriteWidth - objects[objects.length - 1].width) / 2

  let startLeft = (maxWidth - totalWidth) / 2
  if (objects.length === 1) {
    startLeft += firstObjectCorrection
  } else {
    startLeft += (firstObjectCorrection + lastObjectCorrection) / 2
  }

  return [objects, startLeft]
}

/**
 * General use case of rendering scattered sprites
 *
 * @param {Array} sprite - id of sprite to draw
 * @param {CanvasRenderingContext2D} ctx
 * @param {Number} width - width of segment to populate, in meters
 * @param {Number} offsetLeft
 * @param {Number} groundLevel - height at which to draw people
 * @param {Number} randSeed - ensures a consistent sequence of objects across
 *    renders
 * @param {Number} minSpacing - left spacing between each object, in meters
 *    (controls density)
 * @param {Number} maxSpacing - maximum right spacing between each object, in
 *    meters (controls density)
 * @param {Number} adjustment - further spacing adjustment value
 * @param {Number} padding - buffer zone at segments sides to avoid drawing in
 * @param {Number} multiplier
 * @param {Number} dpi
 */
export function drawScatteredSprites (
  sprites,
  ctx,
  width,
  offsetLeft,
  groundLevel,
  randSeed,
  minSpacing = DEFAULT_SCATTER_SPACING_MIN,
  maxSpacing = DEFAULT_SCATTER_SPACING_MAX,
  adjustment = 0,
  padding = 0,
  multiplier,
  dpi
) {
  // Pool must be an array of objects. If an array of strings
  // is passed in, look up its sprite information from the string
  // id and convert to object.
  const pool = sprites.map((sprite) => {
    if (typeof sprite === 'string') {
      const svg = images.get(sprite)
      return {
        id: sprite,
        width: svg.width / TILE_SIZE_ACTUAL
      }
    }

    // Pass objects through as-is
    return sprite
  })

  const maxSpriteWidth = getSpriteMaxWidth(pool)
  const [objects, startLeft] = getRandomObjects(
    pool,
    width,
    randSeed,
    minSpacing,
    maxSpacing,
    maxSpriteWidth,
    adjustment,
    padding
  )

  for (const object of objects) {
    const id = object.id
    const sprite = getSpriteDef(id)
    const svg = images.get(id)

    const distanceFromGround =
      multiplier *
      TILE_SIZE *
      ((svg.height - (sprite.originY ?? object.originY ?? 0)) /
        TILE_SIZE_ACTUAL)

    drawSegmentImage(
      id,
      ctx,
      undefined,
      undefined,
      undefined,
      undefined,
      offsetLeft +
        (object.left -
          // Adjust offset according to the svg viewbox width
          svg.width / TILE_SIZE_ACTUAL / 2 -
          // Adjust according to sprite's defined "hitbox" width
          (maxSpriteWidth - object.width) / 2 +
          startLeft) *
          TILE_SIZE *
          multiplier,
      groundLevel - distanceFromGround,
      undefined,
      undefined,
      multiplier,
      dpi
    )
  }
}
