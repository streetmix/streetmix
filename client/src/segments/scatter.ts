import seedrandom from 'seedrandom'

import { images } from '../app/load_resources'
import { maxBy } from '../util/maxBy'
import { drawSegmentImage } from './view'
import { getSpriteDef } from './info'
import { TILE_SIZE, TILE_SIZE_ACTUAL } from './constants'

const DEFAULT_SELECTION_WEIGHT = 50
const DEFAULT_SCATTER_SPACING_MIN = 0
const DEFAULT_SCATTER_SPACING_MAX = 1 // in meters

interface ScatterableEntity {
  // `id` is a unique identifier. For drawing segments, this `id`
  // corresponds to the sprite's "filename", but that association is not
  // particularly meaningful for picking objects.
  id: string
  width: number

  // (optional) affects the rarity of this entity. Higher values mean it has a
  // higher chance of being selected, relative to other entities in the same
  // pool. Lower numbers are rarer. If not provided, an entity's default weight
  // is 50.
  weight?: number

  // (optional) if `true`, prevent this entity from being selected first in a
  // list (reserved for unique or special entities that should not be the only
  // object displayed if we only have space for one object.)
  disallowFirst?: boolean

  // Don't provide this value; it's added by `getRandomObjects()`
  left?: number

  // Allow extending this with arbitrary properties
  [x: string]: unknown
}

type EntityPool = ScatterableEntity[]

/**
 * Given a pool of entities, with defined widths, get the maximum width value
 * This number is used during rendering to adjust sprites for center alignment
 */
function getSpriteMaxWidth (pool: EntityPool): number {
  return maxBy(pool, (s) => s.width).width
}

/**
 * Using the random generator function, perform a weighted random selection
 * from a pool of entities, then clone and return the object for that entity.
 */
function pickRandomEntityFromPool (
  pool: EntityPool,
  randomGenerator: seedrandom.PRNG
): ScatterableEntity {
  const totalWeight = pool.reduce((sum, entity) => {
    return sum + (entity.weight ?? DEFAULT_SELECTION_WEIGHT)
  }, 0)
  const randomNumber = Math.floor(randomGenerator() * totalWeight)

  let pickedEntityIndex = 0
  let searchWeight = 0

  for (let i = 0; i < pool.length; i++) {
    const entity = pool[i]
    searchWeight += entity.weight ?? DEFAULT_SELECTION_WEIGHT

    if (randomNumber < searchWeight) {
      pickedEntityIndex = i
      break
    }
  }

  // Clone the entity and return it
  return Object.assign({}, pool[pickedEntityIndex])
}

/**
 * Given a pool of entities, and a few parameters, draw continuously from that
 * pool until it has fully populated the given segment width. Because it uses
 * segment width to know when to stop drawing, this function also does double
 * duty by setting distances between objects.
 *
 * This function is exported for unit testing only.
 */
export function getRandomObjects (
  pool: EntityPool,
  maxWidth: number, // width of segment, in meters
  randSeed: string = '9123984', // default randSeed if one is not provided
  minSpacing: number, // in meters (controls density)
  maxSpacing: number, // in meters (controls density)
  maxSpriteWidth: number, // maximum sprite width in the pool (via `getSpriteMaxWidth()`)
  spacingAdjustment: number = 0, // additional value to adjust spacing, in meters
  padding: number = 0 // buffer zone at segment sides
): [EntityPool, number] {
  // Initialize a random generator function with `randSeed` here. Don't
  // recreate it and definitely not within loops, which is not performant.
  const randomGenerator = seedrandom(randSeed)

  const objects = []
  let runningWidth = 0
  let previousEntityId = null
  let currentEntityId = null
  let lastWidth = 0

  // Pick (draw) objects from the pool until we have at least one object, and
  // until we hit the maximum width given. `padding` is used to so that
  // we don't render objects too closely to the edge of a segment.
  while (objects.length === 0 || runningWidth < maxWidth - padding * 2) {
    let object: ScatterableEntity

    // Special choosing logic only for larger pools.
    if (pool.length >= 4) {
      do {
        object = pickRandomEntityFromPool(pool, randomGenerator)
        currentEntityId = object.id
      } while (
        // Don't pick the same object twice in a row, and avoid picking
        // certain objects as the first object in a list. This is because
        // some objects are "special" and look strange when on their own,
        // but can diversify when included in a list.
        currentEntityId === previousEntityId ||
        (objects.length === 0 && object.disallowFirst === true)
      )
    } else {
      object = pickRandomEntityFromPool(pool, randomGenerator)
      currentEntityId = object.id
    }

    previousEntityId = currentEntityId
    object.left = runningWidth

    // Calculate the amount of space to allocate to this object,
    // creating space for placing the next object (if any).
    // First, take into account the object's defined sprite width.
    // Then add `minSpacing`, the minimum spacing between sprites.
    // Next, add a variable distance between sprites. This number
    // is randomly chosen between `minSpacing` and `maxSpacing.`
    // The `randomGenerator()` must be used to ensure that this random
    // number is consistent across renders.
    // Lastly, this is further tweaked by adding a `spacingAdjustment`
    // value, which can adjust overall density. In the case of people,
    // this value makes sprites slightly denser, to account for the
    // transparent space around each person sprite's defined width.
    lastWidth =
      object.width +
      minSpacing +
      randomGenerator() * (maxSpacing - minSpacing) +
      spacingAdjustment
    runningWidth += lastWidth

    objects.push(object)
  }

  // After exiting the loop, remove the space allocated for the next object,
  // by undoing the addition of `lastWidth` to `runningWidth`.
  // TODO: Refactor this behavior.
  const totalWidth = runningWidth - lastWidth

  // The left position of sprites are along the left edge.
  // This correction adjusts the actual left render position, based on the
  // width of the first and last sprites, to better balance centering the
  // sprites within the segment.
  const firstObjectCorrection = (maxSpriteWidth - objects[0].width) / 2
  const lastObjectCorrection =
    (maxSpriteWidth - objects[objects.length - 1].width) / 2

  let startLeft = (maxWidth - totalWidth) / 2
  if (objects.length === 1) {
    startLeft += firstObjectCorrection
  } else {
    startLeft += (firstObjectCorrection + lastObjectCorrection) / 2
  }

  return [objects, startLeft]
}

/**
 * General use case of rendering scattered sprites
 */
export function drawScatteredSprites (
  sprites: EntityPool,
  ctx: CanvasRenderingContext2D,
  width: number,
  offsetLeft: number,
  groundLevel: number,
  randSeed: string,
  minSpacing: number = DEFAULT_SCATTER_SPACING_MIN,
  maxSpacing: number = DEFAULT_SCATTER_SPACING_MAX,
  adjustment: number = 0,
  padding: number = 0,
  multiplier: number,
  dpi: number
): void {
  // Pool must be an array of objects. If an array of strings
  // is passed in, look up its sprite information from the string
  // id and convert to object.
  const pool: EntityPool = sprites.map((sprite): ScatterableEntity => {
    if (typeof sprite === 'string') {
      const svg = images.get(sprite)
      return {
        id: sprite,
        width: svg.width / TILE_SIZE_ACTUAL
      }
    }

    // Pass objects through as-is
    return sprite
  })

  const maxSpriteWidth = getSpriteMaxWidth(pool)
  const [objects, startLeft] = getRandomObjects(
    pool,
    width,
    randSeed,
    minSpacing,
    maxSpacing,
    maxSpriteWidth,
    adjustment,
    padding
  )

  for (const object of objects) {
    const id = object.id
    const sprite = getSpriteDef(id)
    const svg = images.get(id)

    const distanceFromGround =
      multiplier *
      TILE_SIZE *
      ((svg.height - (sprite.originY ?? object.originY ?? 0)) /
        TILE_SIZE_ACTUAL)

    drawSegmentImage(
      id,
      ctx,
      undefined,
      undefined,
      undefined,
      undefined,
      offsetLeft +
        ((object.left ?? 0) -
          // Adjust offset according to the svg viewbox width
          svg.width / TILE_SIZE_ACTUAL / 2 -
          // Adjust according to sprite's defined "hitbox" width
          (maxSpriteWidth - object.width) / 2 +
          startLeft) *
          TILE_SIZE *
          multiplier,
      groundLevel - distanceFromGround,
      undefined,
      undefined,
      multiplier,
      dpi
    )
  }
}
