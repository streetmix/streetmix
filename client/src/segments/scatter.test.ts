import { vi } from 'vitest'

import { getRandomObjects } from './scatter'
import PEOPLE from './people.json'

// Provide mock people data to prevent changes in production data from
// breaking the expected values of this test
vi.mock('./people.json', () => ({
  default: require('./__mocks__/people.json')
}))

// The unit test for `getRandomObjects()` is designed so that the expected
// return values are very precise decimal numbers, on purpose. We want to see
// the same values on each test, so that refactoring it does not cause new
// regressions. We want the actual appearance of streets to remain as stable
// as possible.
//
// However, using these exact numbers can be brittle, because there are
// certain unavoidable cases where the resulting values will be different.
// This usually happens when the random number generator gets called with
// different input, for example, when the number of people in the `people.json`
// pool have changed. We have mocked this data, so changing the mock data
// is an expected cause for breaking the test and needing this to update.
//
// Another reason is if the random number generator is called in a slightly
// different order because we have changed the logic. This is an example that
// requires justification to break the test. Is changing the logic necessary
// to make a performance improvement, or add a new feature? If so, then this
// can cause a change in output that we expect, and we should update the test
// to allow those values to change. But let's say we refactored the code to
// make use of a new JavaScript language feature, and we do not expect the
// output to be different. If it were (and tests fail), we can then go back
// and see whether a mistake was made.
describe('scatter objects in segments', () => {
  it('picks people at normal density', () => {
    const [people, startLeft] = getRandomObjects(
      PEOPLE,
      28.66667,
      'randseed',
      0,
      5,
      4,
      -0.5,
      2
    )

    expect(people).toMatchSnapshot()
    expect(startLeft).toEqual(4.396675322336476)
  })

  it('picks people at sparse density', () => {
    const [people, startLeft] = getRandomObjects(
      PEOPLE,
      28.66667,
      'randseed',
      3.5,
      11.8333,
      4,
      -0.5,
      2
    )

    expect(people).toMatchSnapshot()
    expect(startLeft).toEqual(4.020026451873447)
  })

  it('picks people at dense density', () => {
    const [people, startLeft] = getRandomObjects(
      PEOPLE,
      28.66667,
      'randseed',
      0,
      1.5,
      4,
      -0.5,
      2
    )

    expect(people).toMatchSnapshot()
    expect(startLeft).toEqual(3.3839283366260755)
  })

  it('picks from a pool of one single string id', () => {
    const scatter = {
      sprites: ['plants--bush'],
      minSpacing: -3,
      maxSpacing: 3,
      padding: 2
    }
    const pool = [
      {
        id: 'plants--bush',
        width: 5
      }
    ]

    const [objects, startLeft] = getRandomObjects(
      pool,
      20,
      'randseed',
      scatter.minSpacing,
      scatter.maxSpacing,
      5,
      0,
      scatter.padding
    )

    expect(objects).toMatchSnapshot()
    expect(startLeft).toEqual(4.174402542959051)
  })
})
