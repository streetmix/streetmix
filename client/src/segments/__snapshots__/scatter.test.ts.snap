// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scatter objects in segments > picks from a pool of one single string id 1`] = `
[
  {
    "id": "plants--bush",
    "left": 0,
    "width": 5,
  },
  {
    "id": "plants--bush",
    "left": 6.424349548541568,
    "width": 5,
  },
  {
    "id": "plants--bush",
    "left": 11.651194914081897,
    "width": 5,
  },
]
`;

exports[`scatter objects in segments > picks people at dense density 1`] = `
[
  {
    "id": "The thicc one",
    "left": 0,
    "width": 3,
  },
  {
    "id": "character-05",
    "left": 3.606087387135392,
    "name": "The popular one",
    "weight": 100,
    "width": 2,
  },
  {
    "id": "The thicc one",
    "left": 5.912798728520474,
    "width": 3,
  },
  {
    "id": "character-02",
    "left": 9.655001618279844,
    "name": "Bob",
    "width": 2,
  },
  {
    "id": "The boisterous one",
    "left": 11.861995806598115,
    "weight": 65,
    "width": 3,
  },
  {
    "id": "character-02",
    "left": 14.916528561094843,
    "name": "Bob",
    "width": 2,
  },
  {
    "id": "The boisterous one",
    "left": 17.25277032282099,
    "weight": 65,
    "width": 3,
  },
  {
    "id": "character-01",
    "left": 20.70040088905788,
    "name": "Ashe",
    "width": 2,
  },
  {
    "disallowFirst": true,
    "id": "character-04",
    "left": 23.39881332674785,
    "name": "The shy one",
    "weight": 20,
    "width": 2,
  },
]
`;

exports[`scatter objects in segments > picks people at normal density 1`] = `
[
  {
    "id": "The thicc one",
    "left": 0,
    "width": 3,
  },
  {
    "id": "character-05",
    "left": 6.186957957117974,
    "name": "The popular one",
    "weight": 100,
    "width": 2,
  },
  {
    "id": "The thicc one",
    "left": 10.375995761734915,
    "width": 3,
  },
  {
    "id": "character-02",
    "left": 17.016672060932816,
    "name": "Bob",
    "width": 2,
  },
  {
    "id": "The boisterous one",
    "left": 20.87331935532705,
    "weight": 65,
    "width": 3,
  },
]
`;

exports[`scatter objects in segments > picks people at sparse density 1`] = `
[
  {
    "id": "The thicc one",
    "left": 0,
    "width": 3,
  },
  {
    "id": "character-05",
    "left": 12.14490534881024,
    "name": "The popular one",
    "weight": 100,
    "width": 2,
  },
  {
    "id": "The thicc one",
    "left": 21.62661709625311,
    "width": 3,
  },
]
`;
